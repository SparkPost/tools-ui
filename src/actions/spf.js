
export function saveHistory(domain, status) {
  return {
    type: 'SPARKPOST_API_REQUEST',
    meta: {
      type: 'SPF_SAVE_HISTORY',
      url: '/messaging-tools/spf/history',
      method: 'post',
      data: { domain, status }
    }
  };
}

export function inspect(domain) {
  return {
    type: 'SPARKPOST_API_REQUEST',
    meta: {
      type: 'SPF_INSPECT',
      url: '/messaging-tools/spf/query',
      params: { domain },
      chain: {
        success: ({ dispatch, getState, results: { status } }) => {
          const { auth } = getState();
          if (auth.loggedIn) {
            dispatch(saveHistory(domain, status));
          }
        }
      }
    }
  };
}

export function expandAll() {
  return { type: 'SPF_TREE_EXPAND_ALL' };
}

export function collapseAll() {
  return { type: 'SPF_TREE_COLLAPSE_ALL' };
}

export function expand(id) {
  return { type: 'SPF_TREE_EXPAND', payload: id };
}

export function collapse(id) {
  return { type: 'SPF_TREE_COLLAPSE', payload: id };
}

export function getHistory() {
  return {
    type: 'SPARKPOST_API_REQUEST',
    meta: {
      type: 'SPF_GET_HISTORY',
      method: 'GET',
      url: '/messaging-tools/spf/history'
    }
  };
}
