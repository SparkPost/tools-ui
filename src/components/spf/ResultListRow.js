import React, { Component } from 'react';
import { Link } from 'react-router';
import Icon from 'components/Icon';

class ResultListRow extends Component {
  render() {
    const { domain, timestamp } = this.props;

    return (
      <div>
        <div>
          {this.renderStatus()}
        </div>
        <div>
          {domain}
          <span>Inspected on {timestamp}</span>
        </div>
        <div>
          <Link to={`/spf/results/${domain}`}>
            Refresh
          </Link>
        </div>
      </div>
  );
  }

  renderStatus() {
    const { status } = this.props;
    let iconCls = '';
    switch (status) {
      case 'valid':
        iconCls = 'check-status';
        break;
      case 'warning':
        iconCls = 'warning';
        break;
      case 'error':
        iconCls = 'exclamation-circle';
        break;
    }
    return (<Icon name={iconCls} />);
  }
}

ResultListRow.propTypes = {
  domain: React.PropTypes.string.isRequired,
  status: React.PropTypes.string.isRequired,
  timestamp: React.PropTypes.string.isRequired
};

export default ResultListRow;
