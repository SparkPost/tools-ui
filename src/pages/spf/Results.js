/* eslint-disable no-console */
import React, { Component } from 'react';
import axios from 'axios';
import moment from 'moment';
import config from 'config/index';
import ResultsHeader from './components/ResultsHeader';
import ResultsErrors from './components/ResultsErrors';
import { BackLink } from 'components/button/Button';
// import SPFTree from './components/SPFTree';

import SPFNode from './components/SPFNode';

export default class Results extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      results: { errors: [], warnings: [], spf_tree: {}}
    };
  }

  componentDidMount() {
    // call the api to get domain results
    return this.getResults(this.props.params.domain);
  }

  getResults(domain) {
    this.setState({loading: true});

    return axios.get(`${config.apiBase}/messaging-tools/spf/query`, {params: { domain }})
      .then(({ data }) => {
        const results = data.results;
        results.spf_tree = walkTree(results.spf_tree);
        console.log('walked tree:', results.spf_tree); // eslint-disable-line no-console
        results.timestamp = moment().format('MMM D YYYY[, at] h:mm A');
        this.setState({ results });
      })
      .catch((err) => {
        // TODO show error
      })
      .then(() => this.setState({loading: false}), () => this.setState({loading: false}));
  }

  renderLoading() {
    return (
      <div>
        <BackLink to='/spf-inspector' title='Back to SPF Inspector' />
        <div className='panel panel--accent'>
          <div className='panel__body text--center paddingTop--xxl paddingBottom--xxl'>
            <h4 className='text--muted'>Inspecting {this.props.params.domain}...</h4>
          </div>
        </div>
      </div>
    );
  }

  render() {
    const { results, loading } = this.state;
    const { domain } = this.props.params;
    const { errors, warnings, spf_tree } = results;

    if (loading) {
      return this.renderLoading();
    }

    return (
      <div>
        <BackLink to='/spf-inspector' title='Back to SPF Inspector' />
        <ResultsHeader results={results} domain={domain} refresh={() => this.getResults(domain)} ></ResultsHeader>
        <ResultsErrors errors={errors} warnings={warnings}></ResultsErrors>
        <SPFNode {...spf_tree} domain={domain}>{spf_tree.children}</SPFNode>
      </div>
    );
  }
}

function walkTree(node, i = '0') {
  const walked = Object.assign({}, node, {
    expanded: false,
    treeId: i
  });

  if (node.children) {
    walked.children = node.children.map((child, j) => walkTree(child, `${i}.${j}`));
  }

  return walked;
}
