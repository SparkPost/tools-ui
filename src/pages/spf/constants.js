import { baseUrl } from 'helpers/getCurrentUrl';

export const INTRO_TEXT = 'Visualize your SPF records. Identify which servers are authorized to send on behalf of a domain.';

export const inspectorMeta = {
  title: 'SPF Inspector',
  meta: [
    {
      name: 'description',
      content: 'SPF email authentication is important, but hard to grok. Check SPF rules for sending from your domain with this easy tool from the email experts at SparkPost.'
    },
    // Og
    {
      property: 'og:title',
      content: 'SPF Inspector'
    },
    {
      property: 'og:url',
      content: 'https://tools.sparkpost.com/spf/inspector'
    },
    {
      property: 'og:description',
      content: 'SPF email authentication is important, but hard to grok. Check SPF rules for sending from your domain with this easy tool from the email experts at SparkPost.'
    },
    {
      property: 'og:image',
      content: `${baseUrl}/images/inspector_og.png`
    },
    // Twitter
    {
      name: 'twitter:title',
      content: 'SPF Inspector'
    },
    {
      name: 'twitter:description',
      content: 'SPF email authentication is important, but hard to grok. Check SPF rules for sending from your domain with this easy tool from the email experts at SparkPost.'
    },
    {
      name: 'twitter:card',
      content: 'summary_large_image'
    },
    {
      name: 'twitter:image',
      content: `${baseUrl}/images/inspector_twitter.png`
    }
  ]
};
