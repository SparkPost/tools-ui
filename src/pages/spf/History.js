import React, { Component } from 'react';
import axios from 'axios';

import ResultListRow from 'components/spf/ResultListRow';

class History extends Component {
  constructor(props) {
    super(props);
    this.state = {
      results: [],
      loading: false
    };
  }

  componentDidMount() {
    this.getResults();
  }

  // TODO move this to redux
  getResults() {
    this.setState({ loading: true });
    return axios({
      method: 'get',
      url: 'http://api.sparkpost.dev/api/v1/messaging-tools/spf/history'
    })
    .then(({ data }) => {
      const { results } = data;
      this.setState({
        results: results.map(({ domain, status, timestamp }, idx) => (
          {
            id: idx, domain, status, timestamp
          }
        )),
        loading: false
      });
    }, ({ response: { data: { errors = [] }}}) => {
      this.setState({
        error: errors[0],
        loading: false
      });
    });
  }

  renderResults() {
    const { results } = this.state;

    if (results.length === 0) {
      return this.renderEmpty();
    }

    return results.map((values) => this.renderResultListRow(values));
  }

  renderResultListRow(row) {
    return (
        <ResultListRow key={row.id} {...row} />
    );
  }

  renderEmpty() {
    return (
        <div className='text--center paddingTop--md'>
          <p className='text--regular text--muted'>Search for SPF records to start saving results!</p>
        </div>
    );
  }

  renderLoading() {
    return (
        <div className='text--center paddingTop--md'>
          <p className='text--regular text--muted'>Loading History...</p>
        </div>
    );
  }

  render() {
    const { loading } = this.state;
    return (
        <div className='flex center-xs'>
          <div className='col-xs-12 col-md-10 col-lg-7'>
            <h1>SPF History</h1>
            {loading ? this.renderLoading() : this.renderResults()}
          </div>
        </div>
    );
  }
}

History.defaultProps = {
  loggedIn: false
};

History.propTypes = {
  loggedIn: React.PropTypes.bool.isRequired
};

export default History;
