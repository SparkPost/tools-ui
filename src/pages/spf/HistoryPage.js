import React, { Component } from 'react';
import axios from 'axios';
import config from 'config/index';

import ResultListRow from './components/ResultListRow';

class HistoryPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      results: [],
      loading: false
    };
  }

  componentDidMount() {
    this.getResults();
  }

  // TODO move this to redux
  getResults() {
    this.setState({ loading: true });
    return axios({
      method: 'get',
      url: `${config.apiBase}/messaging-tools/spf/history`
    })
    .then(({ data }) => {
      const { results } = data;
      this.setState({
        results: results.map(({ domain, status, timestamp }, idx) => (
          {
            id: idx, domain, status, timestamp
          }
        )),
        loading: false
      });
    })
    .catch((err) => {
      this.setState({
        error: err,
        loading: false
      });
    });
  }

  renderResults() {
    const { results } = this.state;

    if (results.length === 0) {
      return (
        <div className='text--center paddingTop--md'>
          <p className='text--regular text--muted'>Search for SPF records to start saving results!</p>
        </div>
      );
    }

    return results.map((row, i) => <ResultListRow key={i} {...row} />);
  }

  renderLoading() {
    return (
        <div className='text--center paddingTop--md'>
          <p className='text--regular text--muted'>Loading History...</p>
        </div>
    );
  }

  render() {
    const { loading } = this.state;
    return (
        <div className='flex center-xs'>
          <div className='col-xs-12 col-md-10 col-lg-7'>
            <h1>SPF History</h1>
            {loading ? this.renderLoading() : this.renderResults()}
          </div>
        </div>
    );
  }
}

HistoryPage.defaultProps = {
  loggedIn: false
};

HistoryPage.propTypes = {
  loggedIn: React.PropTypes.bool.isRequired
};

export default HistoryPage;
