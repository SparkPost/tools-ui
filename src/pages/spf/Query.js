/* eslint-disable no-console */

import React, { Component } from 'react';
import { withRouter } from 'react-router';
import classNames from 'classnames';
import { INTRO_TEXT } from './constants';
import { LinkButton } from 'components/button/Button';


class Query extends Component {
  constructor(props) {
    super(props);
    this.state = {
      domain: '',
      domainError: false
    };
  }

  validateDomain() {
    // does this sorta kinda look like a domain
    return this.state.domain && this.state.domain.indexOf('.') > 0;
  }

  componentDidMount() {
    this.props.router.setRouteLeaveHook(this.props.route, () => {
      const validDomain = this.validateDomain(this.state.domain);
      this.setState({domainError: !validDomain});
      return validDomain;
    });
  }

  renderPanel() {
    return (
      <div className='panel panel--accent text--left'>
        <div className='panel__body'>
          <h4>Inspect a Domain</h4>
          <p className='paddingBottom--md text--muted'>Enter a domain to inspect, e.g., sparkpost.com.</p>
          { this.renderForm() }
        </div>
      </div>
    );
  }

  renderForm() {
    return (
      <form>
        <div className="flex">

          <div className={classNames('col-xs-8', {'has-error': this.state.domainError })}>
            <input className='input__text input--full' type='text' onChange={(e) => this.setState({domain: e.target.value}) } placeholder='mydomain.com' />
            <span className="input__error">Please enter a valid domain.</span>
          </div>

          <div className='col-xs-4'>
            <LinkButton type='orange' fullWidth={true} to={ `/spf/results/${this.state.domain}` }>View Results</LinkButton>
          </div>

        </div>
      </form>
    );
  }

  render() {
    return (
      <div className='flex center-xs'>
        <div className='col-xs-12 col-md-10 col-lg-7'>
          <h1>SPF Inspector</h1>
          <p className='marginBottom--lg text--muted'>{INTRO_TEXT}</p>
          { this.renderPanel() }
        </div>
      </div>
    );
  }
}

export default withRouter(Query);
