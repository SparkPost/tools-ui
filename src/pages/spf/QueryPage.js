import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import classNames from 'classnames';

import { INTRO_TEXT } from './constants';
import HistoryList from './components/HistoryList';

export class QueryPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      domain: '',
      domainError: false
    };
  }

  validateDomain() {
    // does this sorta kinda look like a domain
    return this.state.domain && this.state.domain.indexOf('.') > 0;
  }

  handleSubmit(e) {
    e.preventDefault();

    const validDomain = this.validateDomain(this.state.domain);
    this.setState({domainError: !validDomain});

    if (validDomain) {
      this.props.router.push(`/spf/inspector/results/${this.state.domain}`);
    }
  }

  renderPanel() {
    return (
      <div className='panel panel--accent text--left'>
        <div className='panel__body'>
          <h4 className='marginBottom--xxs'>Inspect a Domain</h4>
          <p className='paddingBottom--md'>Enter a domain to inspect, e.g., sparkpost.com.</p>
          { this.renderForm() }
        </div>
      </div>
    );
  }

  renderForm() {
    return (
      <form onSubmit={ (e) => this.handleSubmit(e) }>
        <div className="flex">

          <div className='col-xs-12 col-md-8'>
            <div className={classNames('input__group', {'has-error': this.state.domainError })}>
              <input className='input__text input--full' type='text' onChange={(e) => this.setState({domain: e.target.value}) } placeholder='mydomain.com' />
              {this.state.domainError && <span className="input__error">Please enter a valid domain.</span>}
            </div>
          </div>

          <div className='col-xs-12 col-md-4 responsiveBump'>
            <button type='submit' className='button button--orange button--full'>View Results</button>
          </div>

        </div>
      </form>
    );
  }

  render() {
    return (
      <div className='flex center-xs'>
        <div className='col-xs-12 col-md-10 col-lg-7'>
          <h1>SPF Inspector</h1>
          <p className='marginBottom--lg'>{INTRO_TEXT}</p>
          { this.renderPanel() }
          { this.props.loggedIn && <HistoryList /> }
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ auth }) => ({ loggedIn: auth.loggedIn });
export default withRouter(connect(mapStateToProps)(QueryPage));
