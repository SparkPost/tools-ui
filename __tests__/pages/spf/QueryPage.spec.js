import React from 'react';
import renderer from 'react-test-renderer';
import { QueryPage } from 'pages/spf/QueryPage';

describe('QueryPage Snapshots', () => {
  test('baseline snapshot', () => {
    expect(renderer.create(<QueryPage loggedIn={false}></QueryPage>)).toMatchSnapshot();
  });
});
