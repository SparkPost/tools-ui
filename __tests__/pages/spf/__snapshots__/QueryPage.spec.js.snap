exports[`QueryPage Snapshots baseline snapshot 1`] = `
<div
  className="flex center-xs">
  <div
    className="col-xs-12 col-md-10 col-lg-7">
    <h1>
      SPF Inspector
    </h1>
    <p
      className="marginBottom--lg">
      Visualize your SPF records. Identify which servers are authorized to send on behalf of a domain.
    </p>
    <div>
      <div
        style={
          Object {
            "background": "#fa6423",
            "color": "#fff",
            "padding": "15px",
          }
        }>
        Sorry, this tool is down for a scheduled maintenance right now. Please check back in a little bit.
      </div>
      <div
        className="panel panel--accent text--left">
        <div
          className="panel__body">
          <h4
            className="marginBottom--xxs">
            Inspect a Domain
          </h4>
          <p
            className="paddingBottom--md">
            Enter a domain to inspect, e.g., sparkpost.com.
          </p>
          <form
            onSubmit={[Function]}>
            <div
              className="flex">
              <div
                className="col-xs-12 col-md-8">
                <div
                  className="input__group">
                  <input
                    className="input__text input--full"
                    onChange={[Function]}
                    placeholder="mydomain.com"
                    type="text" />
                </div>
              </div>
              <div
                className="col-xs-12 col-md-4 responsiveBump">
                <button
                  className="button button--orange button--full"
                  type="submit">
                  View Results
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
